colcon_cargo/task/cargo/test.py

```diff
@@ -31,29 +38,98 @@ async def test(self, *, additional_hooks=None):  # noqa: D102
+            cargo_args = []
+
         # invoke cargo test
-        rc = await run(
+        unit_rc = await run(
             self.context,
-            [CARGO_EXECUTABLE, 'test', '-q',
-                '--target-dir', test_results_path],
-            cwd=args.path, env=env)
+            self._test_cmd(cargo_args),
+            cwd=args.path, env=env, capture_output=True)
+
+        fmt_rc = await run(
+            self.context,
+            self._fmt_cmd(),
+            cwd=args.path, env=env, capture_output=True)
+
+        error_report = self._create_error_report(unit_rc, fmt_rc)
+        with open(test_results_path, 'wb') as result_file:
+            xmlstr = minidom.parseString(eTree.tostring(error_report))
+            xmlstr = xmlstr.toprettyxml(indent='    ', encoding='utf-8')
+            result_file.write(xmlstr)
 
-        if rc.returncode:
+        if unit_rc.returncode or fmt_rc.returncode:
             self.context.put_event_into_queue(TestFailure(pkg.name))
             # the return code should still be 0
         return 0
+
+    def _test_cmd(self, cargo_args):
+        args = self.context.args
+        return [
+            CARGO_EXECUTABLE,
+            'test',
+            '--quiet',
+            '--target-dir',
+            args.build_base,
+        ] + cargo_args + [
+            '--',
+            '--color=never',
+        ]
+
+    # Ignore cargo args for rustfmt
+    def _fmt_cmd(self):
+        return [
+            CARGO_EXECUTABLE,
+            'fmt',
+            '--check',
+            '--',
+            '--color=never',
+        ]
+
+    def _create_error_report(self, unit_rc, fmt_rc) -> eTree.Element:
+        # TODO(luca) revisit when programmatic output from cargo test is
+        # stabilized, for now just have a suite for unit, and fmt tests
+        failures = 0
+        testsuites = eTree.Element('testsuites')
+        # TODO(luca) add time
+        testsuite = eTree.SubElement(testsuites,
+                                     'testsuite', {'name': 'cargo_test'})
+        unit_testcase = eTree.SubElement(testsuite, 'testcase',
+                                         {'name': 'unit'})
+        if unit_rc.returncode:
+            unit_failure = eTree.SubElement(unit_testcase, 'failure',
+                                            {'message': 'cargo test failed'})
+            unit_failure.text = unit_rc.stdout.decode('utf-8')
+            failures += 1
+        fmt_testcase = eTree.SubElement(testsuite, 'testcase', {'name': 'fmt'})
+        if fmt_rc.returncode:
+            fmt_failure = eTree.SubElement(fmt_testcase, 'failure',
+                                           {'message': 'cargo fmt failed'})
+            fmt_failure.text = fmt_rc.stdout.decode('utf-8')
+            failures += 1
+        testsuite.attrib['errors'] = str(0)
+        testsuite.attrib['failures'] = str(failures)
+        testsuite.attrib['skipped'] = str(0)
+        testsuite.attrib['tests'] = str(2)
```

Comment:
Perhaps I'm misunderstanding, but it seems like we're hard coding the report to always list 2 tests run?

Reply:
Yes that is the case, all the output from `cargo test` is collated into a single result, and the result of `cargo fmt --check` is into another result. This is due to it not being possible right now to separate the result of `cargo test` by test case (unless we migrate to nextest or try to parse human readable output. It's part of the "what's next" in the PR description (Test granularity)